File: server/internal/service/transaction.go
package service

import (
	"context"

	"cashcompass-server/internal/model"
	"cashcompass-server/internal/repository"
)

type TransactionService interface {
	List(ctx context.Context, filter model.TransactionFilter) ([]model.Transaction, error)
	GetByID(ctx context.Context, id int) (model.Transaction, error)
	Create(ctx context.Context, p model.CreateTransactionParams) (model.Transaction, error)
	UpdateAmount(ctx context.Context, id int, amount int) error
	UpdateLabel(ctx context.Context, id int, label string) error
	UpdateAccount(ctx context.Context, id int, accountID *int, accountLabel string) error
	UpdateCategory(ctx context.Context, id int, categoryID *int, categoryLabel string) error
	UpdateDate(ctx context.Context, id int, iso8601, date string) error
	Delete(ctx context.Context, id int) error
	SumByMonth(ctx context.Context, accountType, dateFrom, dateTo string) ([]model.MonthSum, error)
	BalancesByMonth(ctx context.Context, accountIDs []int) ([]model.AccountMonthBalance, error)
}

type TransactionServiceImpl struct {
	transactions repository.TransactionRepository
}

func NewTransactionService(repo repository.TransactionRepository) *TransactionServiceImpl {
	return &TransactionServiceImpl{transactions: repo}
}

func (s *TransactionServiceImpl) List(ctx context.Context, filter model.TransactionFilter) ([]model.Transaction, error) {
	return s.transactions.List(ctx, filter)
}

func (s *TransactionServiceImpl) GetByID(ctx context.Context, id int) (model.Transaction, error) {
	return s.transactions.GetByID(ctx, id)
}

func (s *TransactionServiceImpl) Create(ctx context.Context, p model.CreateTransactionParams) (model.Transaction, error) {
	return s.transactions.Create(ctx, p)
}

func (s *TransactionServiceImpl) UpdateAmount(ctx context.Context, id int, amount int) error {
	return s.transactions.UpdateAmount(ctx, id, amount)
}

func (s *TransactionServiceImpl) UpdateLabel(ctx context.Context, id int, label string) error {
	return s.transactions.UpdateLabel(ctx, id, label)
}

func (s *TransactionServiceImpl) UpdateAccount(ctx context.Context, id int, accountID *int, accountLabel string) error {
	return s.transactions.UpdateAccount(ctx, id, accountID, accountLabel)
}

func (s *TransactionServiceImpl) UpdateCategory(ctx context.Context, id int, categoryID *int, categoryLabel string) error {
	return s.transactions.UpdateCategory(ctx, id, categoryID, categoryLabel)
}

func (s *TransactionServiceImpl) UpdateDate(ctx context.Context, id int, iso8601, date string) error {
	return s.transactions.UpdateDate(ctx, id, iso8601, date)
}

func (s *TransactionServiceImpl) Delete(ctx context.Context, id int) error {
	return s.transactions.Delete(ctx, id)
}

func (s *TransactionServiceImpl) SumByMonth(ctx context.Context, accountType, dateFrom, dateTo string) ([]model.MonthSum, error) {
	return s.transactions.SumByMonth(ctx, accountType, dateFrom, dateTo)
}

func (s *TransactionServiceImpl) BalancesByMonth(ctx context.Context, accountIDs []int) ([]model.AccountMonthBalance, error) {
	return s.transactions.BalancesByMonth(ctx, accountIDs)
}

File: src/services/transaction.py
from __future__ import annotations
from datetime import datetime
from typing import Optional
from src.models import Transaction, TransactionFilter, MonthSum, AccountMonthBalance
from src.repository.transaction import TransactionRepository


class TransactionService:
    def __init__(self, repo: TransactionRepository):
        self.repo = repo

    @staticmethod
    def _normalize_label(label: str) -> str:
        normalized = label.strip()
        if not normalized:
            raise ValueError("transaction label is required")
        return normalized

    @staticmethod
    def _normalize_amount(amount: int) -> int:
        if amount == 0:
            raise ValueError("transaction amount must be non-zero")
        return amount

    @staticmethod
    def _normalize_date(iso8601: str, date: str) -> tuple[str, str]:
        iso_date = (iso8601 or "").strip()[:10]
        storage_date = (date or "").strip()
        if not iso_date or not storage_date:
            raise ValueError("transaction date is required")
        try:
            datetime.strptime(iso_date, "%Y-%m-%d")
        except ValueError as exc:
            raise ValueError("transaction date must be YYYY-MM-DD") from exc
        if storage_date != iso_date.replace("-", "_"):
            raise ValueError("transaction date is invalid")
        return iso_date, storage_date

    @staticmethod
    def _require_account(account_id: Optional[int], account_label: str) -> tuple[int, str]:
        if account_id is None or not account_label.strip():
            raise ValueError("transaction account is required")
        return account_id, account_label.strip()

    def list(self, f: TransactionFilter) -> list[Transaction]:
        return self.repo.list(f)

    def get_by_id(self, id: int) -> Transaction:
        return self.repo.get_by_id(id)

    def create(self, iso8601: str, date: str, amount: int, label: str,
               account_id: Optional[int], account_label: str,
               category_id: Optional[int], category_label: str) -> Transaction:
        iso8601, date = self._normalize_date(iso8601, date)
        label = self._normalize_label(label)
        amount = self._normalize_amount(amount)
        account_id, account_label = self._require_account(account_id, account_label)
        return self.repo.create(iso8601, date, amount, label, account_id, account_label, category_id, category_label)

    def update_amount(self, id: int, amount: int):
        self.repo.update_amount(id, self._normalize_amount(amount))

    def update_label(self, id: int, label: str):
        self.repo.update_label(id, self._normalize_label(label))

    def update_account(self, id: int, account_id: Optional[int], account_label: str):
        account_id, account_label = self._require_account(account_id, account_label)
        self.repo.update_account(id, account_id, account_label)

    def update_category(self, id: int, category_id: Optional[int], category_label: str):
        self.repo.update_category(id, category_id, category_label)

    def update_date(self, id: int, iso8601: str, date: str):
        iso8601, date = self._normalize_date(iso8601, date)
        self.repo.update_date(id, iso8601, date)

    def delete(self, id: int):
        self.repo.delete(id)

    def sum_by_month(self, account_type: str, date_from: str = "", date_to: str = "") -> list[MonthSum]:
        return self.repo.sum_by_month(account_type, date_from, date_to)

    def balances_by_month(self, account_ids: list[int]) -> list[AccountMonthBalance]:
        return self.repo.balances_by_month(account_ids)
